keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,577 @@
+* The disk may be degrading which affects its performance.
+
+If the volume is mounted on a pod and used by an application, the following problems could also happen.
+* There may be read/write I/O errors.
+* The file system on the volume may be corrupted.
+* Filesystem may be out of capacity.
+* Volume may be unmounted by accident outside of Kubernetes.
+
+If the CSI driver has implemented the CSI volume health function proposed in this design document, Kubernetes could communicate with the CSI driver to retrieve any errors detected by the underlying storage system. Kubernetes can store this data on the associated PVC so that user can inquire this information and decide how to handle them. For example, if the volume is out of capacity, user can request a volume expansion to get more space. Kubernetes may also use volume health information stored on the PVC to automatically reconcile.
+
+There could be conditions that cannot be reported by a CSI driver. There could be network failure where Kubernetes may not be able to get response from the CSI driver. In this case, a call to the CSI driver may time out. There could be network congestion which causes slow response. One or more nodes where the volume is attached to may be down. This can be monitored and detected by the volume health controller so that user knows what has happened.
+
+The Kubernetes components that may modify the PVC status with volume health information include the following:
+
+* An external monitoring agent on each kubernetes worker node.
+* An external monitoring controller on the master node.
+
+Details will be described in the following proposal section.
+
+## Proposal
+
+Volume monitoring is the main focus of this proposal. Reactions are not in the scope of this proposal.
+
+- If a volume provisioned by the CSI driver is deleted, we need to mark the corresponding PVC to inform users.
+
+- If mounting error occurs, we need to mark the PVC.
+
+- If other errors occur, we also need to mark the PVC.
+
+The main architecture is shown below:
+
+![pv health monitor architecture](./pv-health-monitor.png)
+
+The following areas will be the focus of this proposal at first:
+
+- The health condition checking of volumes themselves. For example, whether the volume is deleted, whether the usage is reaching the threshold, and so on.
+- Mounting conditions checking.
+- Other errors that could affect the usability of the volume.
+
+Three main parts are involved here in the architecture.
+
+- API changes: The Status field in the PVC will be used to mark volumes if they are unhealthy.
+
+- External Controller:
+  - The external controller will be deployed as a sidecar together with the CSI controller driver, similar to how the external-provisioner sidecar is deployed.
+  - Trigger controller RPC to check the health condition of network volumes themselves for network storage.
+  - Watch for node failure events for both network and local storage.
+
+- External Agent:
```

Comment:
So I think the main thing we need to resolve here is:

Should this be a node-level sidecar, or kubelet itself? There's been some effort to push monitoring out of kubelet, but for our purposes, we need to work with [node authorizer](https://github.com/kubernetes/kubernetes/blob/bbc035a66cacc345dd54676e6c4e1f0389dbcb5a/plugin/pkg/auth/authorizer/node/node_authorizer.go#L102) since we want to update the status.

Reply:
Discussed this with @gnufied .  So the current proposal is to have a node-level sidecar update the PVC status. If we do it in Kubelet, as you said we may not be allowed to add more monitoring logic there.  So there is a tradeoff.

@liggitt Hi Jordan, can you help answer the question?  Thanks!

Reply:
Feedback from Jordan: He is concerned about giving a node agent permissions to modify PVC Status.  He asked “what the impact is if a malicious node put an unhealthy status on a PVC”.  His suggestion is to “start with something like reporting events”.

Reply:
Discussed with Michelle.  Removed API changes from the main proposal.  Both controller and node agent will only report events now so that they are consistent.

Reply:
Thanks!

Regarding naming, it may be more clear to call it "external node agent" to easily distinguish it from the external controller